pkg/kubectl/cmd/version.go

```diff
@@ -50,73 +58,91 @@ func NewCmdVersion(f cmdutil.Factory, out io.Writer) *cobra.Command {
 		Long:    "Print the client and server version information for the current context",
 		Example: versionExample,
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunVersion(f, out, cmd)
-			cmdutil.CheckErr(err)
+			options := new(VersionOptions)
+			cmdutil.CheckErr(options.Complete(cmd))
+			cmdutil.CheckErr(options.Validate())
+			cmdutil.CheckErr(options.Run(f, out))
 		},
 	}
 	cmd.Flags().BoolP("client", "c", false, "Client version only (no server required).")
 	cmd.Flags().BoolP("short", "", false, "Print just the version number.")
-	cmd.Flags().String("output", "", "output format, options available are yaml and json")
+	cmd.Flags().String("output", "", "one of 'yaml' or 'json'")
 	cmd.Flags().MarkShorthandDeprecated("client", "please use --client instead.")
 	return cmd
 }
 
-func RunVersion(f cmdutil.Factory, out io.Writer, cmd *cobra.Command) error {
-	var serverVersion *apimachineryversion.Info = nil
-	var serverErr error = nil
-	vo := Version{nil, nil}
+func retrieveServerVersion(f cmdutil.Factory) (*apimachineryversion.Info, error) {
+	discoveryClient, err := f.DiscoveryClient()
+	if err != nil {
+		return nil, err
+	}
+
+	// Always request fresh data from the server
+	discoveryClient.Invalidate()
+	return discoveryClient.ServerVersion()
+}
+
+func (o *VersionOptions) Run(f cmdutil.Factory, out io.Writer) error {
+	var (
+		serverVersion *apimachineryversion.Info
+		serverErr     error
+		versionInfo   Version
+	)
 
 	clientVersion := version.Get()
-	vo.ClientVersion = &clientVersion
+	versionInfo.ClientVersion = &clientVersion
 
-	if !cmdutil.GetFlagBool(cmd, "client") {
+	if !o.clientOnly {
 		serverVersion, serverErr = retrieveServerVersion(f)
-		vo.ServerVersion = serverVersion
+		versionInfo.ServerVersion = serverVersion
 	}
 
-	switch of := cmdutil.GetFlagString(cmd, "output"); of {
+	switch o.output {
 	case "":
-		if cmdutil.GetFlagBool(cmd, "short") {
+		if o.short {
 			fmt.Fprintf(out, "Client Version: %s\n", clientVersion.GitVersion)
-
 			if serverVersion != nil {
 				fmt.Fprintf(out, "Server Version: %s\n", serverVersion.GitVersion)
 			}
 		} else {
 			fmt.Fprintf(out, "Client Version: %s\n", fmt.Sprintf("%#v", clientVersion))
-
 			if serverVersion != nil {
 				fmt.Fprintf(out, "Server Version: %s\n", fmt.Sprintf("%#v", *serverVersion))
 			}
 		}
 	case "yaml":
-		y, err := yaml.Marshal(&vo)
+		marshalled, err := yaml.Marshal(&versionInfo)
 		if err != nil {
 			return err
 		}
-
-		fmt.Fprintln(out, string(y))
+		fmt.Fprintln(out, string(marshalled))
 	case "json":
-		y, err := json.Marshal(&vo)
+		marshalled, err := json.Marshal(&versionInfo)
 		if err != nil {
 			return err
 		}
-		fmt.Fprintln(out, string(y))
+		fmt.Fprintln(out, string(marshalled))
 	default:
-		return errors.New("invalid output format: " + of)
-
+		panic(fmt.Sprintf(
```

Comment:
kubectl should not panic in any cases. It's OK to return an error, even though it may be caused by a bug.